src/ossl/ecc.rs

```diff
@@ -709,3 +714,153 @@ impl Verify for EccOperation {
+        CKD_SHA224_KDF | CKD_SHA224_KDF_SP800 => {
+            OSSL_DIGEST_NAME_SHA2_224.as_ptr()
+        }
+        CKD_SHA256_KDF | CKD_SHA256_KDF_SP800 => {
+            OSSL_DIGEST_NAME_SHA2_256.as_ptr()
+        }
+        CKD_SHA384_KDF | CKD_SHA384_KDF_SP800 => {
+            OSSL_DIGEST_NAME_SHA2_384.as_ptr()
+        }
+        CKD_SHA512_KDF | CKD_SHA512_KDF_SP800 => {
+            OSSL_DIGEST_NAME_SHA2_512.as_ptr()
+        }
+        CKD_SHA3_224_KDF | CKD_SHA3_224_KDF_SP800 => {
+            OSSL_DIGEST_NAME_SHA3_224.as_ptr()
+        }
+        CKD_SHA3_256_KDF | CKD_SHA3_256_KDF_SP800 => {
+            OSSL_DIGEST_NAME_SHA3_256.as_ptr()
+        }
+        CKD_SHA3_384_KDF | CKD_SHA3_384_KDF_SP800 => {
+            OSSL_DIGEST_NAME_SHA3_384.as_ptr()
+        }
+        CKD_SHA3_512_KDF | CKD_SHA3_512_KDF_SP800 => {
+            OSSL_DIGEST_NAME_SHA3_512.as_ptr()
+        }
+        _ => std::ptr::null(),
+    }) as *const c_char
+}
+
+pub const NONE_NAME: &[u8; 1] = b"\0";
+fn ckd_to_type_name(mech: CK_EC_KDF_TYPE) -> *const c_char {
+    (match mech {
+        CKD_SHA1_KDF | CKD_SHA224_KDF | CKD_SHA256_KDF | CKD_SHA384_KDF
+        | CKD_SHA512_KDF | CKD_SHA3_224_KDF | CKD_SHA3_256_KDF
+        | CKD_SHA3_384_KDF | CKD_SHA3_512_KDF => OSSL_KDF_NAME_X963KDF.as_ptr(),
+        _ => NONE_NAME.as_ptr(),
+    }) as *const c_char
+}
+
+fn new_pkey_ctx_from_obj(pkey: &mut EvpPkey) -> KResult<EvpPkeyCtx> {
+    Ok(EvpPkeyCtx::from_ptr(unsafe {
+        EVP_PKEY_CTX_new_from_pkey(
+            get_libctx(),
+            pkey.as_mut_ptr(),
+            std::ptr::null(),
+        )
+    })?)
+}
+
+impl Derive for ECDHOperation<'_> {
+    fn derive(
+        &mut self,
+        key: &Object,
+        template: &[CK_ATTRIBUTE],
+        _mechanisms: &Mechanisms,
+        objfactories: &ObjectFactories,
+    ) -> KResult<(Object, usize)> {
+        if self.finalized {
+            return err_rv!(CKR_OPERATION_NOT_INITIALIZED);
+        }
+        self.finalized = true;
+
+        let mut params = OsslParam::with_capacity(6).set_zeroize().add_int(
+            name_as_char(OSSL_EXCHANGE_PARAM_EC_ECDH_COFACTOR_MODE),
+            if self.cofactor { 1 } else { -1 } as i32,
+        )?;
+        if self.kdf != CKD_NULL {
+            params = params
+                .add_const_c_string(
+                    name_as_char(OSSL_EXCHANGE_PARAM_KDF_TYPE),
+                    ckd_to_type_name(self.kdf),
+                )?
+                .add_const_c_string(
+                    name_as_char(OSSL_EXCHANGE_PARAM_KDF_DIGEST),
+                    ckd_to_digest_name(self.kdf),
+                )?;
+        }
+        if self.shared.len() > 0 {
+            params = params.add_octet_string(
+                name_as_char(OSSL_EXCHANGE_PARAM_KDF_UKM),
+                &self.shared.to_vec(),
+            )?;
+        }
+
+        let mut obj = objfactories.derive_key_from_template(key, template)?;
+        match obj.get_attr_as_ulong(CKA_VALUE_LEN) {
```

Comment:
After some discussion internally sounds like this check is not up to par.
Probably something like  https://github.com/latchset/kryoptic/blob/eee73043e7232714f17c857e034e1bb65de8acaa/src/kdf.rs#L778 should be done here.

Some key type (like AES) already check that CKA_VALUE_LEN is acceptable but only if non 0: https://github.com/latchset/kryoptic/blob/eee73043e7232714f17c857e034e1bb65de8acaa/src/aes.rs#L128

Reply:
Agreed. Will add the check.

Reply:
This should be resolved and similar check was implemented also in the ECDH case.